data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/SqsWorker.java

```diff
@@ -204,22 +214,37 @@ private List<DeleteMessageBatchRequestEntry> processS3EventNotificationRecords(f
 
         LOG.info("Received {} messages from SQS. Processing {} messages.", s3EventNotificationRecords.size(), parsedMessagesToRead.size());
 
+        List<DeleteMessageBatchRequestEntry> waitingForAcknowledgements = new ArrayList<>();
+        AcknowledgementSet acknowledgementSet = null;
+        if (endToEndAcknowledgementsEnabled) {
```

Comment:
This puts all of the messages into a single acknowledgement set. This could lead to bad acknowledgements.

I think the fundamental unit of acknowledgements for S3 is an object. Once an object is fully read it can be queued for deletion.

I'd probably have a callback on S3 which responds back to the SQS worker. The worker then queues up the delete requests per object until it reaches 10 (or perhaps the max message count for receives). Then it performs a batch delete.

Reply:
Let's discuss. I think this code the keeps the current granularity of deletion.